src/core/src/index/lca_db.rs

```diff
@@ -0,0 +1,387 @@
+        let rank = LcaDB::c_char_to_string(tup.rank());
+        LineagePair { 
+            name: name, 
+            rank: rank,
+        }
+    }
+}
+
+impl LineagePair {
+    pub fn new(name: String, rank: String) -> LineagePair {
+        LineagePair {
+            name: name,
+            rank: rank,
+        }
+    }
+    pub fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    pub fn rank(&self) -> String {
+        self.rank.clone()
+    }
+}
+
+#[cfg_attr(all(target_arch = "wasm32", target_vendor = "unknown"), wasm_bindgen)]
+#[derive(Debug, TypedBuilder)]
+pub struct LcaDB {
+    ksize: u32,
+    scaled: u64,
+    moltype: String,
+
+    #[builder(default = "filename".to_string())]
+    filename: String,
+
+    #[builder(default = 0u32)]
+    _next_index: u32,
+
+    #[builder(default = 0u32)]
+    _next_lid: u32,
+
+    #[builder(default = HashMap::new())]
+    ident_to_name: HashMap<String, String>,
+
+    #[builder(default = HashMap::new())]
+    ident_to_idx: HashMap<String, u32>,
+
+    #[builder(default = HashMap::new())]
+    idx_to_lid: HashMap<u32, u32>,
+
+    #[builder(default = HashMap::new())]
+    lineage_to_lid: HashMap<Vec<LineagePair>, u32>,
+
+    #[builder(default = HashMap::new())]
+    lid_to_lineage: HashMap<u32, Vec<LineagePair>>,
+
+    #[builder(default = HashMap::new())]
+    hashval_to_idx: HashMap<u64, Vec<u32>>, //HashSet as keys
+}
+
+impl Default for LcaDB {
+    fn default() -> LcaDB {
+        LcaDB {
+            ksize: 21,
+            scaled: 0,
+            filename: "filename".to_string(),
+            moltype: "DNA".to_string(),
+
+            _next_index: 0,
+            _next_lid: 0,
+            ident_to_name: HashMap::new(),
+            ident_to_idx: HashMap::new(),
+            idx_to_lid: HashMap::new(),
+            lineage_to_lid: HashMap::new(),
+            lid_to_lineage: HashMap::new(),
+            hashval_to_idx: HashMap::new(),
+        }
+    }
+}
+
+impl LcaDB {
+    pub fn new(
+        ksize: u32,
+        scaled: u64,
+        filename: &[u8],
+        moltype: &[u8],
+    ) -> LcaDB {
+
+        let filename = str::from_utf8(filename).unwrap().to_string();
+        let moltype = str::from_utf8(moltype).unwrap().to_string();
+        
+        let _next_index = 0;
+        let _next_lid = 0;
+        let ident_to_name = HashMap::new();
+        let ident_to_idx = HashMap::new();
+        let idx_to_lid = HashMap::new();
+        let lineage_to_lid = HashMap::new();
+        let lid_to_lineage = HashMap::new();
+        let hashval_to_idx = HashMap::new();
+
+        LcaDB {
+            ksize,
+            scaled,
+            filename,
+            moltype,
+            _next_index,
+            _next_lid,
+            ident_to_name,
+            ident_to_idx,
+            idx_to_lid,
+            lineage_to_lid,
+            lid_to_lineage,
+            hashval_to_idx,
+        }
+    }
+
+    pub fn c_char_to_string(char_ptr: *const c_char) -> String {
```

Comment:
This can be a free function (not really part of `LcaDB`, as it doesn't take `self` as first parameter).

(and this conversion should be happening in `crate::ffi` too =])